File: 680-ValidPalindromeII/680-ValidPalindromeII.java
/*

680. Valid Palindrome II
Easy

Given a non-empty string s, you may delete at most one character. Judge whether you can make it a palindrome.

Example 1:

Input: "aba"
Output: True

Example 2:

Input: "abca"
Output: True
Explanation: You could delete the character 'c'.

Note:

    The string will only contain lowercase characters a-z. The maximum length of the string is 50000.

*/

class Solution {
    public boolean validPalindrome(String s) {
        boolean canSkip = true;
        for(int i = 0, j = s.length() - 1; j - i >= 1; ){
            if(s.charAt(i) == s.charAt(j)){
                i++;
                j--;
                continue;
            }
            else
                return isPalindrome(i + 1, j, s) || isPalindrome(i, j - 1, s);
        }
        return true;
    }
    private boolean isPalindrome(int i, int j, String s){
        while(j - i >= 1){
            if(s.charAt(i) != s.charAt(j))
                return false;
            i++;
            j--;
        }
        return true;
        
    }
}

File: 680-ValidPalindromeII/680-ValidPalindromeII.py
"""

680. Valid Palindrome II

Given a string s, return true if the s can be palindrome after deleting at most one character from it.

Example 1:

Input: s = "aba"
Output: true

Example 2:

Input: s = "abca"
Output: true
Explanation: You could delete the character 'c'.

Example 3:

Input: s = "abc"
Output: false

Constraints:

    1 <= s.length <= 105
    s consists of lowercase English letters.

"""

class Solution:
    def validPalindrome(self, s: str) -> bool:
        def checkPalin(lo, hi):
            while lo < hi: 
                if s[lo] != s[hi]: return False
                lo += 1
                hi -= 1
            return True

        lo, hi, count = 0, len(s) - 1, 0
        while lo < hi:
            if s[lo] == s[hi]:
                lo += 1
                hi -= 1
                continue
            else : return checkPalin(lo + 1, hi) or checkPalin(lo, hi - 1)
        return True
